### PR TITLE
AE-962: Contract Variable Monitor agent template

### PR DESCRIPTION
**Pull request by arbitrary-sparrow**
This PR adds a new agent template that can monitor contract variables for percent changes in value over time. The following sample `agent-config.json` can be used to facilitate running the unit tests:
```json
{
  "developerAbbreviation": "AE",
  "protocolName": "Uniswap",
  "protocolAbbreviation": "UNI",
  "contracts": {
    "UniswapV3Pool": {
      "address": "0x8ad599c3A0ff1De082011EFDDc58f1908eb6e6D8",
      "abiFile": "UniswapV3Pool.json",
      "variables": {
        "liquidity": {
          "type": "Info",
          "severity": "Low",
          "upperThresholdPercent": 15,
          "lowerThresholdPercent": 5,
          "numDataPoints": 3
        }
      }
    }
  }
}
```
Additionally, here is a simplified `UniswapV3Pool.json`:
```json
{
  "abi": [
    {
      "inputs": [],
      "name": "liquidity",
      "outputs": [
        {
          "internalType": "uint128",
          "name": "",
          "type": "uint128"
        }
      ],
      "stateMutability": "view",
      "type": "function"
    }
  ]
}
```

### CODE REVIEW

File: contract-variable-monitor/SETUP.md [hunk 1]
```diff
@@ -0,0 +1,199 @@
+# Contract Variable Monitor Agent Template
+
+This agent monitors contract variable values for specified contract addresses. Alert type and severity
+are specified per variable per contract address.
+
+## Agent Setup Walkthrough
+
+The following steps will take you from a completely blank template to a functional agent.
+
+1. Open the `agent-config.json` file.
```
Comment (echoAE) [inline contract-variable-monitor/SETUP.md:10]: I like the idea of committing `agent-config.json.example` to the template repository, then instructing the developer to copy it to `agent-config.json` when they populate it.

File: contract-variable-monitor/SETUP.md [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+          "lowerThresholdPercent": 5,
+          "numDataPoints": 1
+        }
+      }
+    }
+  }
+```
+
+In this example, all functions should be called against the proxy Unitroller, but the definitions for
+the functions that belong to the Comptroller contract will subsequently be a part of the Comproller
```
Comment (echoAE) [inline contract-variable-monitor/SETUP.md:92]:
```suggestion
the functions that belong to the Comptroller contract will subsequently be a part of the Comptroller
```

File: contract-variable-monitor/package.json [hunk 3]
```diff
@@ -0,0 +1,32 @@
+{
+  "name": "forta-{developerAbbreviation}-{protocolAbbreviation}-admin-events",
```
Comment (echoAE) [inline contract-variable-monitor/package.json:2]:
```suggestion
  "name": "forta-{developerAbbreviation}-{protocolAbbreviation}-variable-monitor",
```

File: contract-variable-monitor/src/utils.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+  // eslint-disable-next-line global-require,import/no-dynamic-require
+  const { abi } = require(`../abi/${abiName}`);
+  return abi;
+}
+
+// get a list of variable getter information objects for each variable name listed for a given
+// contract in the config
+function getVariableInfo(contractConfig, currentContract, variableList, contractList) {
+  const proxyName = contractConfig.proxy;
+  const { variables } = contractConfig;
```
Comment (echoAE) [inline contract-variable-monitor/src/utils.js:14]:
```suggestion
  const { variables, proxy: proxyName } = contractConfig;
```

File: contract-variable-monitor/SETUP.md [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+6. If a contract is a proxy for another contract, where the ABI for getter function for a variable will
+be located in the proxy contract's ABI JSON file, the entry in the `agent-config.json` file may look like
+the following:
+
+```json
+  "contracts": {
+    "Unitroller": {
+      "address": "0x3d9819210A31b4961b30EF54bE2aeD79B9c9Cd3B",
+      "abiFile": "Unitroller.json",
+      "proxy": "NonfungibleTokenPositionDescriptor"
```
Comment (echoAE) [inline contract-variable-monitor/SETUP.md:74]:
```suggestion
      "proxy": "Comptroller"
```

File: contract-variable-monitor/src/utils.js [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    const proxyVariables = variableList[proxyName].variables;
+    if (proxyVariables) {
+      if (variableList === undefined) {
+        // eslint-disable-next-line no-param-reassign
+        variableList = { ...proxyVariables };
+      } else {
+        // eslint-disable-next-line no-param-reassign
+        variableList = { ...variableList, ...proxyVariables };
+      }
+    }
```
Comment (echoAE) [inline contract-variable-monitor/src/utils.js:38]: I'm uncertain what this is intended to do.  If line 29 succeeds, there's no way that the conditional on line 31 will evaluate to `true`.

Also, the variable `variableList` is not used anywhere else in this function after line 36.  Is the intention to update the `configEntries` variable in the calling function (from `agent.js`)?

It looks like line 29 should remain and lines 30-38 should be removed.

File: contract-variable-monitor/src/utils.js [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+function checkThreshold(thresholdPercent, currValue, pastValues) {
+  const thresholdPercentBN = new BigNumber(thresholdPercent);
+  const averageBN = pastValues.getAverage();
+  const differenceBN = currValue.minus(averageBN).abs();
+  const differencePercentBN = differenceBN.div(averageBN).times(100);
+  let percentOver;
+
+  if (differencePercentBN.gt(thresholdPercentBN)) {
+    percentOver = differencePercentBN;
+    return percentOver;
```
Comment (echoAE) [inline contract-variable-monitor/src/utils.js:124]:
```suggestion
```

File: README.md [hunk 8]
```diff
@@ -32,4 +32,9 @@ addresses. Alert type and severity are specified per function per contract addre
...
 
 This agent monitors blockchain transactions that have failed and are associated with a specific
-contract address. Alert type and severity are both configurable.
\ No newline at end of file
+contract address. Alert type and severity are both configurable.
+
+### Contract Variable Monitor
+
+This agent monitors contract variable values for specified contract addresses. Alert type and severity
+are specified per variable per contract address.
```
Comment (echoAE) [inline README.md:40]: Perhaps a brief mention of the percent thresholds in the description for this agent template.

File: contract-variable-monitor/src/agent.spec.js [hunk 9]
```diff
@@ -0,0 +1,431 @@
...
+      // the call to getAbi will fail if the file does not exist
+      const abi = utils.getAbi(abiFile);
+
+      // get all of the function objects from the loaded ABI file
+      const functionObjects = getObjectsFromAbi(abi, 'function');
+
+      // for all of the variable names specified, verify that their corresponding getter function
+      // exists in the ABI
+      Object.keys(variables).forEach((variableName) => {
+        expect(Object.keys(functionObjects).indexOf(variableName)).not.toBe(-1);
```
Comment (echoAE) [inline contract-variable-monitor/src/agent.spec.js:73]: Do we want to check that the output variable type is numeric?

File: contract-variable-monitor/src/agent.spec.js [hunk 10]
```diff
@@ -0,0 +1,431 @@
...
+      // upperThresholdPercent change
+      const newValue = initialGetterValue + 1;
+      mockContract[functionInConfig.name] = jest.fn().mockResolvedValue(newValue);
+
+      // run the agent again now that we have seen the minimum number of data points
+      const findings = await handleBlock();
+      expect(findings).toStrictEqual([]);
+    });
+
+    it('returns a finding when the value decreases and the change is less than the lower threshold percent change', async () => {
```
Comment (echoAE) [inline contract-variable-monitor/src/agent.spec.js:346]:
```suggestion
    it('returns a finding when the value decreases more than the lower threshold percent change', async () => {
```

**Review by echoAE — COMMENTED** [review]
Mostly minor updates suggested.

File: contract-variable-monitor/src/utils.js (hunk 6 above)
Comment (arbitrary-sparrow) [reply contract-variable-monitor/src/utils.js]: I removed the proxy logic entirely in favor of adding an extra note in `SETUP.md` for using proxy contracts

File: contract-variable-monitor/src/agent.spec.js (hunk 9 above)
Comment (arbitrary-sparrow) [reply contract-variable-monitor/src/agent.spec.js:73]: Good point

File: contract-variable-monitor/src/utils.js [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+function getAbi(abiName) {
+  // eslint-disable-next-line global-require,import/no-dynamic-require
+  const { abi } = require(`../abi/${abiName}`);
+  return abi;
+}
+
+// get a list of variable getter information objects for each variable name listed for a given
+// contract in the config
+function getVariableInfo(contractConfig, currentContract, variableList) {
```
Comment (echoAE) [inline contract-variable-monitor/src/utils.js:12]:
```suggestion
function getVariableInfo(contractConfig, currentContract) {
```

File: contract-variable-monitor/src/utils.js [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+  const { variables, proxy: proxyName } = contractConfig;
```
Comment (echoAE) [inline contract-variable-monitor/src/utils.js:13]:
```suggestion
  const { variables } = contractConfig;
```

**Review by echoAE — APPROVED** [review]
A few proxy remnants to remove.

**Review by arbitrary-CodeBeholder — APPROVED** [review]

**Pull request merged**